File: statem/schema.py
import inspect
import logging
from collections.abc import Awaitable, Callable
from dataclasses import dataclass, field
from functools import cached_property
from typing import Any, Generic, Literal, NamedTuple, TypeAlias, TypeVar

from pydantic import BaseModel, ConfigDict, Field, field_validator

T = TypeVar("T")

logger = logging.getLogger(__name__)


class GuardError(TypeError):
    """Raised when a guard function returns a non-bool value."""


class TransitionError(Exception):
    """Raised when an action fails during a transition."""


# — Runtime primitives ———————————————————————————————————————————————————


@dataclass(frozen=True, slots=True)
class Signal:
    """A signal that triggers a transition.

    Attributes:
        event: Event name (upper-case by convention, e.g. `"START"`).
        data: Arbitrary payload dict; defaults to empty.
    """

    event: str
    data: dict[str, Any] = field(default_factory=dict)


HookSource: TypeAlias = Literal["on", "always", "entry", "exit"]


class ResultEntry(NamedTuple):
    """A single recorded guard or action result, in execution order.

    Attributes:
        state: Name of the state the machine was in when this fired.
        source: Lifecycle hook: `"on"`, `"always"`, `"entry"`, or `"exit"`.
        kind: `"guard"` or `"action"`.
        name: Registered name of the guard or action function.
        value: `bool` for guards; return value (may be `None`) for actions.
    """

    state: str
    source: HookSource
    kind: Literal["guard", "action"]
    name: str
    value: Any


@dataclass(slots=True, kw_only=True)
class Context(Generic[T]):
    """Short-lived execution context passed to every action and guard during one `run()` call.

    Created at the start of `StateMachine.run` and discarded when it returns. The machine updates
    `current_state` as transitions fire. All fields are keyword-only.

    Generic over `T`, the type of `session`. `Context` used bare (unparameterized) behaves exactly
    as before -- `session` types as `Any`. Guards/actions that want typed access to their own
    session shape can annotate their `ctx` parameter as `Context[BankSession]` (or whatever their
    session type is) to get full type-checking and autocomplete on `ctx.session`.

    Attributes:
        run_id: Correlation id for this `run()` call, used in log lines. `None` if the caller
            didn't supply one -- `StateMachine.run()` is responsible for generating one (a
            `uuid4().hex` string) before constructing this object; `Context` itself just stores
            whatever it's given.
        current_state: Name of the state the machine is currently in; updated by the engine as
            each transition fires.
        session: Caller-owned, opaque payload of any shape; the engine never inspects it, only
            threads it through to actions/guards.
        history: Ordered list of every state entered during this `run()` call; the first entry
            is always the initial state.
        results: Ordered list of `ResultEntry` for every guard and action executed, in the exact
            order they fired during this `run()` call.
    """

    run_id: str | None = None
    current_state: str
    session: T
    history: list[str] = field(init=False)
    results: list[ResultEntry] = field(init=False, default_factory=list)

    def __post_init__(self) -> None:
        self.history = [self.current_state]


# — Callable type aliases ————————————————————————————————————————————————

ActionFn: TypeAlias = Callable[[Context, Signal], Awaitable[Any] | Any]

GuardFn: TypeAlias = Callable[[Context, Signal], bool] | Callable[[Context, Signal], Awaitable[bool]]


# — Registries ——————————————————————————————————————————————————————————


class _RegEntry(NamedTuple):
    """Internal registry entry: pairs a callable with its async flag, cached at registration time."""

    fn: Callable[..., Any]
    is_async: bool


class ActionRegistry:
    """Registry of named action functions (sync or async)."""

    __slots__ = ("_fns",)

    def __init__(self) -> None:
        self._fns: dict[str, _RegEntry] = {}

    def register(self, name: str, fn: ActionFn) -> None:
        """Register a single named action function (sync or async)."""
        self._fns[name] = _RegEntry(fn=fn, is_async=inspect.iscoroutinefunction(fn))

    def register_many(self, actions: dict[str, ActionFn]) -> None:
        """Register multiple named action functions at once."""
        for name, fn in actions.items():
            self.register(name, fn)

    def has(self, name: str) -> bool:
        """Return `True` if an action named `name` is registered."""
        return name in self._fns

    def __contains__(self, name: str) -> bool:
        return name in self._fns

    def __len__(self) -> int:
        return len(self._fns)

    async def execute(self, name: str, ctx: Context, signal: Signal, source: HookSource) -> None:
        """Execute a single named action. Raises `KeyError` if not registered."""
        if name not in self._fns:
            raise KeyError(f"'{name}' not registered")
        entry = self._fns[name]
        value = await entry.fn(ctx, signal) if entry.is_async else entry.fn(ctx, signal)
        ctx.results.append(ResultEntry(state=ctx.current_state, source=source, kind="action", name=name, value=value))

    async def execute_many(
        self,
        names: list[str],
        ctx: Context,
        signal: Signal,
        source: HookSource,
    ) -> None:
        """Execute actions in order, awaiting each one."""
        for name in names:
            await self.execute(name, ctx, signal, source)


class GuardRegistry:
    """Registry of named guard functions (sync or async).

    Async detection is cached at register time.
    """

    __slots__ = ("_fns",)

    def __init__(self) -> None:
        self._fns: dict[str, _RegEntry] = {}

    def register(self, name: str, fn: GuardFn) -> None:
        """Register a single named guard function (sync or async)."""
        self._fns[name] = _RegEntry(fn=fn, is_async=inspect.iscoroutinefunction(fn))

    def register_many(self, guards: dict[str, GuardFn]) -> None:
        """Register multiple named guard functions at once."""
        for name, fn in guards.items():
            self.register(name, fn)

    def has(self, name: str) -> bool:
        """Return `True` if a guard named `name` is registered."""
        return name in self._fns

    def __contains__(self, name: str) -> bool:
        return name in self._fns

    def __len__(self) -> int:
        return len(self._fns)

    async def evaluate(self, name: str | None, ctx: Context, signal: Signal, source: HookSource) -> bool:
        """Evaluate a guard. Returns `True` if `name` is `None` (no guard = pass)."""
        if name is None:
            return True
        if name not in self._fns:
            raise KeyError(f"'{name}' not registered")
        entry = self._fns[name]
        raw = await entry.fn(ctx, signal) if entry.is_async else entry.fn(ctx, signal)
        if not isinstance(raw, bool):
            raise GuardError(f"Guard '{name}' must return bool, got {type(raw).__name__!r}")
        ctx.results.append(ResultEntry(state=ctx.current_state, source=source, kind="guard", name=name, value=raw))
        logger.debug("Guard '%s' -> %s", name, raw)
        return raw


# — Config models ————————————————————————————————————————————————————————


class TransitionConfig(BaseModel):
    """Configuration for a single transition candidate.

    Attributes:
        target: Name of the state to transition to.
        guard: Registered guard name that must pass for this transition to fire; `None` means
            the transition always passes.
        actions: Ordered list of registered action names to execute when this transition fires.
    """

    model_config = ConfigDict(frozen=True)

    target: str
    guard: str | None = Field(default=None)
    actions: list[str] = Field(default_factory=list)


class StateConfig(BaseModel):
    """Configuration for one state.

    Extra fields (e.g. `role`, `intent`, `render`) are accepted and silently ignored, so raw
    application config dicts can be passed directly. Each `on` value is normalized to a list of
    `TransitionConfig` at parse time.

    Attributes:
        on: Maps event names to transition candidates.
        always: Eventless transitions checked after every state entry.
        entry: Action names to fire when entering this state.
        exit: Action names to fire when leaving this state.
        error_state: Fallback state on unhandled action errors.
    """

    model_config = ConfigDict(frozen=True, extra="ignore")

    on: dict[str, list[TransitionConfig]] = Field(default_factory=dict)
    always: list[TransitionConfig] = Field(default_factory=list)
    entry: list[str] = Field(default_factory=list)
    exit: list[str] = Field(default_factory=list)
    error_state: str | None = None

    @field_validator("on", mode="before")
    @classmethod
    def _normalize_on(cls, v: Any) -> Any:
        """Normalize each event value to a list of TransitionConfig-compatible dicts."""
        if not isinstance(v, dict):
            return v
        result: dict[str, list[Any]] = {}
        for event, spec in v.items():
            if isinstance(spec, str):
                result[event] = [{"target": spec}]
            elif isinstance(spec, dict):
                result[event] = [spec]
            elif isinstance(spec, list):
                result[event] = [{"target": item} if isinstance(item, str) else item for item in spec]
            else:
                result[event] = spec
        return result

    @field_validator("always", mode="before")
    @classmethod
    def _normalize_always(cls, v: Any) -> Any:
        """Normalize each always item to a TransitionConfig-compatible dict."""
        if not isinstance(v, list):
            return v
        return [{"target": item} if isinstance(item, str) else item for item in v]

    @cached_property
    def available_events(self) -> list[str]:
        """Return the event names this state can receive, in declaration order.

        Excludes the `"*"` wildcard entry -- use `accepts_wildcard` to check whether the state
        has a catch-all handler.

        Example:
            ```python
            cfg = StateConfig.model_validate({
                "on": {"START": "running", "CANCEL": "idle", "*": "error"}
            })
            cfg.available_events  # ["START", "CANCEL"]
            cfg.accepts_wildcard  # True
            ```
        """
        return [name for name in self.on if name != "*"]

    @property
    def accepts_wildcard(self) -> bool:
        """Return `True` if this state has a `"*"` catch-all transition."""
        return "*" in self.on

File: statem-rs/src/schema.rs
//! Core types: signals, execution context, config models, and errors.
//!
//! Mirrors `statem/schema.py` in the Python package. Config shorthand normalization (a
//! transition candidate written as a bare string, a single dict, or a list of either) is
//! replicated via `serde`'s `#[serde(untagged)]` + `#[serde(from = "...")]` idioms instead of
//! Python's `field_validator(mode="before")`.

use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, Ordering};
use std::time::{SystemTime, UNIX_EPOCH};

use indexmap::IndexMap;
use serde::Deserialize;

/// Generate a `run_id`: a nanosecond timestamp combined with a monotonic in-process counter,
/// formatted as hex. Not a crypto-strength unique id (unlike Python's `uuid4`) -- just good
/// enough for log correlation, and dependency-free (see the crate root's dependency note on why
/// `uuid` was dropped).
pub(crate) fn generate_run_id() -> String {
    static COUNTER: AtomicU64 = AtomicU64::new(0);
    let count = COUNTER.fetch_add(1, Ordering::Relaxed);
    let nanos = SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_nanos()).unwrap_or(0);
    format!("{nanos:x}-{count:x}")
}

/// A signal that triggers a transition.
#[derive(Debug, Clone)]
pub struct Signal {
    /// Event name (upper-case by convention, e.g. `"START"`).
    pub event: String,
    /// Arbitrary payload; defaults to empty.
    pub data: HashMap<String, serde_json::Value>,
}

impl Signal {
    pub fn new(event: impl Into<String>) -> Self {
        Signal { event: event.into(), data: HashMap::new() }
    }

    pub fn with_data(event: impl Into<String>, data: HashMap<String, serde_json::Value>) -> Self {
        Signal { event: event.into(), data }
    }
}

/// Which lifecycle hook produced a [`ResultEntry`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HookSource {
    On,
    Always,
    Entry,
    Exit,
}

/// What kind of check/effect a [`ResultEntry`] records.
///
/// Unlike the Python port, a successful action's return value isn't recorded here (see the
/// design notes in the crate root) -- only whether a guard passed/failed, or that an action ran.
/// A *failed* action never gets a `ResultEntry` at all: it aborts the transition attempt
/// immediately, mirroring the Python engine (a failed action's `ResultEntry.append` call never
/// runs since the exception propagates first).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ResultKind {
    Guard(bool),
    Action,
}

/// A single recorded guard check or action run, in execution order.
#[derive(Debug, Clone)]
pub struct ResultEntry {
    pub state: String,
    pub source: HookSource,
    pub name: String,
    pub kind: ResultKind,
}

/// Derives the value broadcast via `STATE_SNAPSHOT`/`STATE_DELTA` from `session`, passed to
/// [`crate::machine::StateMachine::stream`].
#[cfg(feature = "agui")]
pub type StateAccessor<T> = std::sync::Arc<dyn Fn(&T) -> serde_json::Value + Send + Sync>;

/// Short-lived execution context passed to every guard and action during one `run()` call.
///
/// Generic over `T`, the session type. Guards receive `&Context<T>` (read-only); actions receive
/// `&mut Context<T>` (may mutate `session` freely) -- enforced by the borrow checker, not just by
/// convention as in the Python port.
#[derive(Clone)]
pub struct Context<T> {
    /// Correlation id for this run, used in log lines. `None` until `StateMachine::run` fills it
    /// in (generates one if the caller didn't supply one).
    pub run_id: Option<String>,
    /// Correlation id for the broader conversation/session this run belongs to (one thread can
    /// span many runs). `None` until `StateMachine::run`/`stream` fills it in, same as `run_id`.
    pub thread_id: Option<String>,
    pub current_state: String,
    pub session: T,
    /// Every state entered during this run, in order; the first entry is always the initial state.
    pub history: Vec<String>,
    pub results: Vec<ResultEntry>,
    /// Internal -- set by `StateMachine::stream` to a sender it emits AG-UI events onto; left
    /// `None` by `StateMachine::run`. Not meant for guards/actions to use directly; check
    /// [`Context::is_stream`] instead of reading this field.
    #[cfg(feature = "agui")]
    pub(crate) event_tx: Option<futures_channel::mpsc::UnboundedSender<crate::agui::AguiEvent>>,
    /// Internal -- set by `StateMachine::stream` from its `state_accessor` argument. Derives the
    /// value broadcast via `STATE_SNAPSHOT`/`STATE_DELTA` from `session`; `None` means the
    /// default `{"current_state": ...}` view is used instead.
    #[cfg(feature = "agui")]
    pub(crate) state_accessor: Option<StateAccessor<T>>,
    /// Internal -- the last value broadcast via `STATE_SNAPSHOT`/`STATE_DELTA`, kept so
    /// `StateMachine::stream` can diff against it to build the next `STATE_DELTA`.
    #[cfg(feature = "agui")]
    pub(crate) state_snapshot: Option<serde_json::Value>,
}

impl<T> Context<T> {
    pub fn new(current_state: impl Into<String>, session: T, run_id: Option<String>, thread_id: Option<String>) -> Self {
        let current_state = current_state.into();
        Context {
            run_id,
            thread_id,
            history: vec![current_state.clone()],
            current_state,
            session,
            results: Vec::new(),
            #[cfg(feature = "agui")]
            event_tx: None,
            #[cfg(feature = "agui")]
            state_accessor: None,
            #[cfg(feature = "agui")]
            state_snapshot: None,
        }
    }

    /// `true` when this context was created by `StateMachine::stream`, `false` for `run`.
    #[cfg(feature = "agui")]
    pub(crate) fn is_stream(&self) -> bool {
        self.event_tx.is_some()
    }
}

/// Hand-written rather than derived: under the `agui` feature, `state_accessor` is a
/// `dyn Fn(&T) -> Value`, which can't implement `Debug` -- printed as presence/absence instead.
impl<T: std::fmt::Debug> std::fmt::Debug for Context<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut s = f.debug_struct("Context");
        s.field("run_id", &self.run_id)
            .field("thread_id", &self.thread_id)
            .field("current_state", &self.current_state)
            .field("session", &self.session)
            .field("history", &self.history)
            .field("results", &self.results);
        #[cfg(feature = "agui")]
        s.field("is_stream", &self.is_stream()).field("state_snapshot", &self.state_snapshot);
        s.finish()
    }
}

/// Configuration for a single transition candidate.
#[derive(Debug, Clone, Deserialize)]
pub struct TransitionConfig {
    /// Name of the state to transition to.
    pub target: String,
    /// Registered guard name that must pass for this candidate to fire; `None` always passes.
    #[serde(default)]
    pub guard: Option<String>,
    /// Ordered action names to run when this candidate fires.
    #[serde(default)]
    pub actions: Vec<String>,
}

/// Deserialization-only shorthand for one transition candidate: a bare target string, or the
/// full `{target, guard, actions}` form.
#[derive(Deserialize)]
#[serde(untagged)]
enum TransitionShorthand {
    Target(String),
    Full(TransitionConfig),
}

impl From<TransitionShorthand> for TransitionConfig {
    fn from(value: TransitionShorthand) -> Self {
        match value {
            TransitionShorthand::Target(target) => TransitionConfig { target, guard: None, actions: Vec::new() },
            TransitionShorthand::Full(config) => config,
        }
    }
}

/// Deserialization-only shorthand for one event's `on` value: a single candidate, or a list of
/// candidates (a guard chain).
#[derive(Deserialize)]
#[serde(untagged)]
enum OnValue {
    Single(TransitionShorthand),
    Many(Vec<TransitionShorthand>),
}

impl From<OnValue> for Vec<TransitionConfig> {
    fn from(value: OnValue) -> Self {
        match value {
            OnValue::Single(candidate) => vec![candidate.into()],
            OnValue::Many(candidates) => candidates.into_iter().map(Into::into).collect(),
        }
    }
}

#[derive(Deserialize)]
struct StateConfigRaw {
    #[serde(default)]
    on: IndexMap<String, OnValue>,
    #[serde(default)]
    always: Vec<TransitionShorthand>,
    #[serde(default)]
    entry: Vec<String>,
    #[serde(default)]
    exit: Vec<String>,
    #[serde(default)]
    error_state: Option<String>,
}

impl From<StateConfigRaw> for StateConfig {
    fn from(raw: StateConfigRaw) -> Self {
        StateConfig {
            on: raw.on.into_iter().map(|(event, value)| (event, value.into())).collect(),
            always: raw.always.into_iter().map(Into::into).collect(),
            entry: raw.entry,
            exit: raw.exit,
            error_state: raw.error_state,
        }
    }
}

/// Configuration for one state.
///
/// Each `on` value and each `always` item is normalized from shorthand at deserialization time
/// (see [`TransitionShorthand`]/[`OnValue`]) -- by the time you hold a `StateConfig`, `on` is
/// always `event -> Vec<TransitionConfig>` and `always` is always `Vec<TransitionConfig>`.
///
/// `on` is an [`IndexMap`], not a `HashMap`: it's iterated for output (error messages, the
/// `diagram` module's edge ordering), where declaration order matters for readability and for
/// getting the same output on every run -- `HashMap`'s iteration order isn't guaranteed. `Signal`'s
/// `data` field, by contrast, is a plain `HashMap`: it's only ever looked up by key, never
/// iterated for display, so there's no reason to pay for order-preservation there.
#[derive(Debug, Clone, Deserialize)]
#[serde(from = "StateConfigRaw")]
pub struct StateConfig {
    pub on: IndexMap<String, Vec<TransitionConfig>>,
    pub always: Vec<TransitionConfig>,
    pub entry: Vec<String>,
    pub exit: Vec<String>,
    pub error_state: Option<String>,
}

impl StateConfig {
    /// Event names this state can receive via `on`, excluding the `"*"` wildcard.
    pub fn available_events(&self) -> Vec<String> {
        self.on.keys().filter(|name| name.as_str() != "*").cloned().collect()
    }

    /// Whether this state has a `"*"` catch-all handler.
    pub fn accepts_wildcard(&self) -> bool {
        self.on.contains_key("*")
    }
}

/// Raised by [`crate::machine::StateMachine::from_config`] when a transition target (in `on`,
/// `always`, or `error_state`) doesn't name a real state.
#[derive(Debug, thiserror::Error)]
#[error("invalid transition targets: {0:?}")]
pub struct BuildError(pub Vec<String>);

/// Raised by [`crate::machine::StateMachine::run`].
///
/// [`RunError::ActionFailed`] from an `on`-transition attempt -- either the firing candidate's
/// own `actions`, or the `entry`/`exit` actions of the state it then enters -- is caught by the
/// state active when the signal was dispatched, if it has `error_state` set (see
/// `StateMachine::run`'s docs). `always`-transition action failures are a deliberate exception:
/// they're never caught, since `always` represents a structural, continuously-re-checked
/// invariant rather than a one-shot response to an external event -- a failure there is a
/// config/logic bug worth surfacing loudly, not something to silently route around.
#[derive(Debug, thiserror::Error)]
pub enum RunError {
    #[error("action '{action}' failed in state '{state}'")]
    ActionFailed {
        state: String,
        action: String,
        #[source]
        source: Box<dyn std::error::Error + Send + Sync>,
    },
    #[error("guard '{0}' is not registered")]
    UnknownGuard(String),
    #[error("action '{0}' is not registered")]
    UnknownAction(String),
    #[error("always-transition loop exceeded 100 hops at '{0}'")]
    AlwaysLoopExceeded(String),
}

/// Raised by [`crate::machine::StateMachine::validate_registries`]: every guard/action name
/// referenced anywhere in `config` (`on`, `always`, `entry`, `exit`) that isn't registered,
/// tagged with where it's referenced (e.g. `"idle.on.START: can_start"`).
///
/// Both lists are always present (possibly empty) rather than conditionally omitted the way the
/// Python original's joined message does -- simpler to construct, same information either way.
#[derive(Debug, thiserror::Error)]
#[error("unregistered guards: {missing_guards:?}, unregistered actions: {missing_actions:?}")]
pub struct ValidationError {
    pub missing_guards: Vec<String>,
    pub missing_actions: Vec<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn generate_run_id_produces_distinct_nonempty_ids() {
        let a = generate_run_id();
        let b = generate_run_id();
        assert!(!a.is_empty());
        assert!(!b.is_empty());
        assert_ne!(a, b);
    }

    #[test]
    fn context_preserves_explicit_run_id() {
        let ctx = Context::new("idle", (), Some("custom-id".to_string()), None);
        assert_eq!(ctx.run_id.as_deref(), Some("custom-id"));
    }

    #[test]
    fn context_leaves_run_id_none_when_omitted() {
        let ctx = Context::new("idle", (), None, None);
        assert_eq!(ctx.run_id, None);
    }

    #[test]
    fn context_preserves_explicit_thread_id() {
        let ctx = Context::new("idle", (), None, Some("custom-thread".to_string()));
        assert_eq!(ctx.thread_id.as_deref(), Some("custom-thread"));
    }
}
